### PR TITLE
fix concurrency on stream execute engine primitives

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR adds the required lock on the engine primitive with Streaming support where a local state is maintained and needs to be synchronized between parallel calls from underlying callbacks.

The primitive were either returning the wrong result or were causing VTGate panics.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/14581

## Checklist

-   [X] "Backport to:" labels have been added if this change should be back-ported
-   [X] Tests were added or are not required
-   [X] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Review by frouioui — APPROVED** [review]

File: go/vt/vtgate/engine/projection.go [hunk 1]
```diff
@@ -88,6 +88,7 @@ func (p *Projection) TryStreamExecute(ctx context.Context, vcursor VCursor, bind
 	env := evalengine.NewExpressionEnv(ctx, bindVars, vcursor)
 	var once sync.Once
 	var fields []*querypb.Field
+	var mu sync.Mutex
```
Comment (dbussink) [inline go/vt/vtgate/engine/projection.go:91]: @harshit-gangal Is this one necessary? There's already locking in the implementation here which should afaik handle any concurrency issues. 

@vmg Afaik the `evalengine.NewExpressionEnv` should be concurrent usage safe, or is it not?

File: go/vt/vtgate/engine/limit.go [hunk 2]
```diff
@@ -108,6 +110,8 @@ func (l *Limit) TryStreamExecute(ctx context.Context, vcursor VCursor, bindVars
 			return nil
 		}
 
+		mu.Lock()
+		defer mu.Unlock()
```
Comment (dbussink) [inline go/vt/vtgate/engine/limit.go:114]: Looks like need this for the count, but wonder, should we do this with an atomic counter instead? The usage might be complicated though, so the lock might be a lot easier for now and then to optimize later.

File: go/vt/vtgate/engine/filter.go [hunk 3]
```diff
@@ -78,9 +79,14 @@ func (f *Filter) TryExecute(ctx context.Context, vcursor VCursor, bindVars map[s
...
 // TryStreamExecute satisfies the Primitive interface.
 func (f *Filter) TryStreamExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	var mu sync.Mutex
+
 	env := evalengine.NewExpressionEnv(ctx, bindVars, vcursor)
 	filter := func(results *sqltypes.Result) error {
 		var rows [][]sqltypes.Value
+
+		mu.Lock()
+		defer mu.Unlock()
```
Comment (dbussink) [inline go/vt/vtgate/engine/filter.go:89]: Similar here, do we guard anything with this one? I think `evalengine.NewExpressionEnv` should be concurrent usage safe.

File: go/vt/vtgate/engine/projection.go (hunk 1 above)
Comment (vmg) [reply go/vt/vtgate/engine/projection.go:91]: No, it definitely isn't, it contains the VM that is used to execute the expressions. You cannot evaluate two expressions at once in the same VM.

Comment (dbussink) [reply go/vt/vtgate/engine/projection.go:91]: Ah ok, so we gotta lock here then as well for now unless we'd want to move creating the env inside the callback then.

File: go/vt/vtgate/engine/filter.go (hunk 3 above)
Comment (dbussink) [reply go/vt/vtgate/engine/filter.go:89]: Turns out, it is not so we should either move it inside or do a lock.

File: go/vt/vtgate/engine/projection.go [hunk 4]
```diff
@@ -88,9 +88,10 @@ func (p *Projection) TryStreamExecute(ctx context.Context, vcursor VCursor, bind
...
 	return vcursor.StreamExecutePrimitive(ctx, p.Input, bindVars, wantfields, func(qr *sqltypes.Result) error {
 		var err error
-		if wantfields {
+		if wantfields && qr.Fields != nil {
```
Comment (dbussink) [inline go/vt/vtgate/engine/projection.go:94]: @harshit-gangal This change breaks a bunch of tests it looks like. Why was this changed?

Comment (dbussink) [reply go/vt/vtgate/engine/projection.go:94]: It looks like this change was made due to the testing in `projection_test.go`, but I suspect it's a problem with that test? Since it runs concurrent stream results, but even if that happens, it's guaranteed that there would always be a field result first, so I don't think the test models a real scenario then.

Comment (dbussink) [reply go/vt/vtgate/engine/projection.go:94]: I also think that now that we lock here to also protect the evalengine expression env, we don't need the more elaborate locking around `wantfields` anymore.

**Review by dbussink — APPROVED** [review]

**Review by vmg — APPROVED** [review]

File: go/vt/vtgate/engine/projection.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/engine/projection.go:91]: The next plan is to create env per shard.

Comment (harshit-gangal) [reply go/vt/vtgate/engine/projection.go:91]: That refactor would not be part of the backport and will be made as a separate PR.
